### PR TITLE
Reduce re-sends and fix hang associated with too many resends

### PR DESCRIPTION
**Pull request by tra**
tl;dr - This should fix a large number of the hangs that we currently experience.

I noticed that many of the hangs that I was studying were involved with re-sending packets that had already been sent.  So I set out to figure out how that worked and improve it.  The first problem area was the computation of the average round-trip times and the deviation of that time.  These values are used to estimate the nextSendTime so if they are too small it will try to resend it too early.  I found that it was using a form of an exponential moving average, which on the surface is a great thing to do!  Unfortunately, the implementation used a lot of integer divides (with bit shifts) and the result of those calculations biased the results towards zero in a bad way.  For example, if you had an average roundTrip of 20 and got a new value of 19, it would adjust the average down to 19.  But it would've taken a value of 52 in order to adjust that average up to 21!  Ultimately, I chose to use floating point calculations and a more rigorous implementation as described in https://fanf2.user.srcf.net/hermes/doc/antiforgery/stats.pdf
Doing this removed a lot of the hocus-pocus and gives us a more explainable way to compute the next send time using statistics to determine a reasonable estimate of how long to wait for a resend.

The 2nd problem with the old implementation was that it used a fixed value, urgentResendTime (defaults to 64ms), to determine when to re-send "urgent" packets.  And since pretty much ALL game-time packets are considered 'urgent', this can lead to a situation where ALL packets could get re-sent on a laggy connection.  I modified this part of the code to use a similarly dynamically estimated value called urgentRetransmitTime.

Finally, and this is the BIG fix, I discovered that the game can be put into an infinite loop in cases where the resends pile up in such a way that it takes longer to process them than the retransmitTime.  This is in the part of the code where it is checking the turboMode flag and calling AsyncWrite().  This is called in a method that is ultimately called as a byproduct of UDPWrite() which creates an infinite loop.  This results in the program just trying to send the same messages, over and over and over again and never stops to read incoming messages.  That in turn causes all of the other clients to enter similar loops.  To fix this, I just made sure it could be only called a fixed number of times (3 for now).  And it seems to prevent this situation from happening.  YAY!

### CODE REVIEW

**Comment by assertivist** [comment]
🥇 **Medal of Avara honor**

**Comment by tra** [comment]
> 🥇 **Medal of Avara honor**

Thank you for this great honor.  I’d like to thank all the packets that were sacrificed to the void of infinite loops.  They had but one life to live and they died so that others could live.  May all their packet brethren reach their destinations and that the grenades that they represent may reach their intended targets (i.e. not me).
